### DISCUSSION TITLE
How can I avoid the assembly runner locking the test DLL?

### DISCUSSION THREAD

**Discussion by crazyfox55** [xunit/xunit#3035] [Question] [answered]
Hello,

I've been working on this issue for a while now. The issue is that when I use the V2 Assembly Runner to run the test project of another repository it will prevent rebuilding the test project.

The flow that I'm going through is this:
1. dotnet build
2. Assembly Runner
3. Make changes to source code
4. dotnet build - Fails here. 
C:\Program Files\dotnet\sdk\8.0.400\Microsoft.Common.CurrentVersion.targets(4857,5): error MSB3021: Unable to copy file
 "obj\Debug\net8.0\Game.Tests.dll" to "bin\Debug\net8.0\Game.Tests.dll". The process cannot access the file '\GameRepo\Game.Tests\bin\Debug\net8.0\Game.Tests.dll' because
 it is being used by another process.
 
 I looked into it deeply and found that the V2 runner loads the dll into the current app domain. In this way it will never be able to release the lock on that dll. I gave up at this point. However V3 runner doesn't load the dll, it could work in my scenario. But I'm still getting the same issue with it locking the dll. 
 
 What can I do? Should I just use the `dotnet test` command instead? I was avoiding this so that I didn't have to parse the output stream. Also V3 is not fully released.
 
 Thank you.

**Comment by bradwilson** [comment, marked as answer]
> What can I do?

There's really only one way to do this: don't run the test assembly where it sits, because loading it locks it. Copy it somewhere else and run that copy instead.

You can have this done automatically for you, but you must:

- Build your tests for .NET Framework (not .NET Core/.NET 5+)
- Make sure app domains are enabled: https://xunit.net/docs/configuration-files#appDomain
- Make sure shadow copy is enabled: https://xunit.net/docs/configuration-files#shadowCopy

If you do all of that, then the .NET Framework will actually do all the heavy lifting of copying your assemblies to a temporary folder and loading them from there, instead of loading them from your build directory.

I'm guessing since your sample had `net8.0` in the path, that you probably can't make your tests run on .NET Framework. That means it'll be up to you to make a copy of the build folder to a temporary place before you run anything. That includes every file and every folder inside the build folder, and it _also_ means if you have any test code which reaches out on the file system, to a place that's relative to where the DLL file lives but above the build folder, that code will be broken, because you're only copying from the build folder down.

You'll also need to clean up the temporary folder when you're done. And you'll need to do this every time before you do any discovery and/or execution, since you want to pick up any newly built binaries in the meantime.

It's not trivial, but it's not that difficult, either.

**Reply by crazyfox55** [reply]
Thank you for the help however I need to run the build test build multiple times it could be 100's of times. So making 100's of copies won't really work. Also there is no clean way to delete the copies once I'm done with that test session because it will still be locked.

Instead I'll just use `dotnet test` and parse the standard output. I think this is my only solution. I don't really need the objects to be directly in my application instead when I parse the output I can create these objects myself.

**Reply by bradwilson** [reply]
> So making 100's of copies won't really work.

I can't help you with that. I don't make the rules. 😂 

> Also there is no clean way to delete the copies once I'm done with that test session because it will still be locked.

Yes, you would have to make your runner a separate executable so that the assemblies would be unlocked after the runner exited.

**Reply by crazyfox55** [reply]
Its not entirely a bad way to go but how would I get the test results back into my main program? I'm not familiar with transferring objects from one process to another, it sounds like a fun challenge.

**Reply by bradwilson** [reply]
There are several ways to do this.

For v3, we chose to send JSON serializations of the messages from the test assembly, for example:

```
$ .\src\xunit.v3.assert.tests\bin\Debug\net6.0\xunit.v3.assert.tests.exe -automated -method BooleanAssertsTests+True.AssertTrue

{"$type":"discovery-starting","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","AssemblyName":"xunit.v3.assert.tests","AssemblyPath":"C:\\Dev\\xunit\\xunit\\src\\xunit.v3.assert.tests\\bin\\Debug\\net6.0\\xunit.v3.assert.tests.dll"}
{"$type":"discovery-complete","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCasesToRun":1}
{"$type":"test-assembly-starting","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","AssemblyName":"xunit.v3.assert.tests","AssemblyPath":"C:\\Dev\\xunit\\xunit\\src\\xunit.v3.assert.tests\\bin\\Debug\\net6.0\\xunit.v3.assert.tests.dll","Seed":1283660447,"StartTime":"2024-10-06T21:49:16.7564359-07:00","TargetFramework":".NETCoreApp,Version=v6.0","TestEnvironment":"64-bit .NET 6.0.33 [collection-per-class, parallel (24 threads)]","TestFrameworkDisplayName":"xUnit.net v3 0.5.0-pre.20-dev+685423ea79"}
{"$type":"test-collection-starting","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestCollectionDisplayName":"Test collection for BooleanAssertsTests+True (id: 151a5052f124ec4da9b6bf0e065698dd682e756bcf1a8c8b94716b86b6b5f895)"}
{"$type":"test-class-starting","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestClassName":"BooleanAssertsTests+True","TestClassSimpleName":"BooleanAssertsTests+True"}
{"$type":"test-method-starting","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestMethodUniqueID":"5e4e37ba9b5d7838df651257acedd7240221067f66cbbe9de0837fadfafaa18c","MethodName":"AssertTrue"}
{"$type":"test-case-starting","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestMethodUniqueID":"5e4e37ba9b5d7838df651257acedd7240221067f66cbbe9de0837fadfafaa18c","TestCaseUniqueID":"e916a9a312a5fb04c415e97610047b7abaae2e2e588a0f512b6941a14fed5a5b","Explicit":false,"TestCaseDisplayName":"BooleanAssertsTests+True.AssertTrue","TestClassMetadataToken":33554534,"TestClassName":"BooleanAssertsTests+True","TestClassSimpleName":"BooleanAssertsTests+True","TestMethodMetadataToken":100663794,"TestMethodName":"AssertTrue","TestMethodParameterTypesVSTest":[],"TestMethodReturnTypeVSTest":"System.Void"}
{"$type":"test-starting","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestMethodUniqueID":"5e4e37ba9b5d7838df651257acedd7240221067f66cbbe9de0837fadfafaa18c","TestCaseUniqueID":"e916a9a312a5fb04c415e97610047b7abaae2e2e588a0f512b6941a14fed5a5b","TestUniqueID":"b2fe2108909ebb09023a101fdab6ee7bb938b0123bc508defd8554bbe1e549c7","Explicit":false,"StartTime":"2024-10-07T04:49:16.7928262+00:00","TestDisplayName":"BooleanAssertsTests+True.AssertTrue","Timeout":0}
{"$type":"test-passed","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestMethodUniqueID":"5e4e37ba9b5d7838df651257acedd7240221067f66cbbe9de0837fadfafaa18c","TestCaseUniqueID":"e916a9a312a5fb04c415e97610047b7abaae2e2e588a0f512b6941a14fed5a5b","TestUniqueID":"b2fe2108909ebb09023a101fdab6ee7bb938b0123bc508defd8554bbe1e549c7","ExecutionTime":0.0064202,"FinishTime":"2024-10-07T04:49:16.8067267+00:00"}
{"$type":"test-finished","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestMethodUniqueID":"5e4e37ba9b5d7838df651257acedd7240221067f66cbbe9de0837fadfafaa18c","TestCaseUniqueID":"e916a9a312a5fb04c415e97610047b7abaae2e2e588a0f512b6941a14fed5a5b","TestUniqueID":"b2fe2108909ebb09023a101fdab6ee7bb938b0123bc508defd8554bbe1e549c7","ExecutionTime":0.0064202,"FinishTime":"2024-10-07T04:49:16.8076834+00:00"}
{"$type":"test-case-finished","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestMethodUniqueID":"5e4e37ba9b5d7838df651257acedd7240221067f66cbbe9de0837fadfafaa18c","TestCaseUniqueID":"e916a9a312a5fb04c415e97610047b7abaae2e2e588a0f512b6941a14fed5a5b","ExecutionTime":0.0064202,"TestsFailed":0,"TestsNotRun":0,"TestsSkipped":0,"TestsTotal":1}
{"$type":"test-method-finished","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","TestMethodUniqueID":"5e4e37ba9b5d7838df651257acedd7240221067f66cbbe9de0837fadfafaa18c","ExecutionTime":0.0064202,"TestsFailed":0,"TestsNotRun":0,"TestsSkipped":0,"TestsTotal":1}
{"$type":"test-class-finished","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","TestClassUniqueID":"ece5da2426b40bd872daeb8d23aa4c23d287c048eae4c092488df994a7b0205b","ExecutionTime":0.0064202,"TestsFailed":0,"TestsNotRun":0,"TestsSkipped":0,"TestsTotal":1}
{"$type":"test-collection-finished","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","TestCollectionUniqueID":"1409f8c2cf27afad6ab4e29cc058ee6c3120813e10e3aa054fd46b6dc2aed421","ExecutionTime":0.0064202,"TestsFailed":0,"TestsNotRun":0,"TestsSkipped":0,"TestsTotal":1}
{"$type":"test-assembly-finished","AssemblyUniqueID":"d14146a22c280b3b6ccc4e366d33938beeb77508d3d3c53f728fb5c35d13bad9","ExecutionTime":0.0573361,"FinishTime":"2024-10-06T21:49:16.8129244-07:00","TestsFailed":0,"TestsNotRun":0,"TestsSkipped":0,"TestsTotal":1}
```

Back in v1/v2, when running across app domains, we used .NET Remoting to pass objects back and forth.
